developers.md

```diff
@@ -1,15 +1,311 @@
+# Developers
 
+- [Introduction](#introduction)
+- [Guide to your first Picker](#guide-to-your-first-picker)
+  - [Requires](#requires)
+  - [First Picker](#first-picker)
+  - [Replacing Actions](#replacing-actions)
+  - [Entry Maker](#entry-maker)
+  - [Oneshot job](#oneshot-job)
+  - [Previewer](#previewer)
+  - [More examples](#more-examples)
+- [Technical](#technical)
+  - [picker](#picker)
+  - [finders](#finders)
+  - [actions](#actions)
+  - [previewers](#previewers)
 
-## Overriding actions/action_set
+## Introduction
+
+So you want to develop your own picker and/or extension for telescope? Then you
+are in the right place! This file will first present an introduction on how to
+do this. After that, this document will present a technical explanation of
+pickers, finders, actions, and the previewer. You will find more information
+in specific help pages and we likely will move some of the technical stuff to
+our vim help docs in the future.
+
+This guide is mainly for telescope so it will assume that a lua knowledge is
+present. You can find information for lua here:
+- [Lua 5.1 Manual](https://www.lua.org/manual/5.1/)
+- [Getting started using Lua in Neovim](https://github.com/nanotee/nvim-lua-guide)
+
+## Guide to your first Picker
+
+To guide you along the way to first picker we will do the following. We will
+open a empty lua scratch file in which we will develop the picker and run it
+each time using `:luafile %`. Later this file then be bundled as extension.
+
+### Requires
+
+The most important includes are the following modules:
+```lua
+local pickers = require "telescope.pickers"
+local finders = require "telescope.finders"
+local conf = require("telescope.config").values
+```
+
+- `pickers` is the main module which is used to create a new picker.
+- `finders` provides interfaces to fill the picker with items.
+- `config` which is used for user configuration and the `values` table holds
+  these configurations. So to make it easier we only get this table in `conf`.
+
+### First Picker
+
+We will now make the most simplest color picker. (Note that the previous snippet
+is also required. We will approach this example step by step).
+
+```lua
+-- our picker function: colors
+local colors = function(opts)
+  pickers.new(opts, {
```

Comment:
out of topic but perhaps we can have here telescope.api that expose confg, finders, actions, pickers at so on.

```lua
-- telescope.api
return setmetatable({}, {
    __index = function(_, module)
      --- 
    end
})
```

```lua
local colors = function(opts)
   local i = require'telescope.api'
  i.pickers.new(opts, {
```

Reply:
hmm. isnt this just replacing require for the sake of replacing require? extension developers should just lazy load the modules they need, if its a big problem.

I dont like this idea.

Reply:
well is for the sake of not having 10 lines + of require statements when creating a custom picker.

Reply:
i am not a big fan of this one module (or metatable fwiw) containing a couple of other modules. That just leads to bad, ambiguous locals. Like `local i` or `local a`. Everybody uses these differently and imo this causes in the best case variable shadowing and in the worst case bugs.

But if you want something like this, you should open an issue and we can talk about it with other people :)

Reply:
Maybe later, though I will keep it in mind